src/analysis/functions.rs

```diff
@@ -238,51 +260,201 @@ fn analyze_function(
     );
     commented |= ret.commented;
 
+    let mut params = func.parameters.clone();
     let mut parameters = function_parameters::analyze(
         env,
-        &func.parameters,
+        &params,
         configured_functions,
         disable_length_detect,
         async,
-        in_trait
+        in_trait,
     );
     parameters.analyze_return(env, &ret.parameter);
 
+    if let Some(ref f) = ret.parameter {
+        if let Type::Function(_) = env.library.type_(f.typ) {
+            if env.config.work_mode.is_normal() {
+                warn!("Function \"{}\" returns callback", func.name);
+                commented = true;
+            }
+        }
+    }
+
     fixup_special_functions(env, imports, name.as_str(), type_tid, &mut parameters);
 
-    for (pos, par) in parameters.c_parameters.iter().enumerate() {
-        assert!(
-            !par.instance_parameter || pos == 0,
-            "Wrong instance parameter in {}",
-            func.c_identifier.as_ref().unwrap()
-        );
-        if let Ok(s) = used_rust_type(env, par.typ, !par.direction.is_out()) {
-            used_types.push(s);
+    // Key: destroy callback index
+    // Value: associated user data index
+    let mut cross_user_data_check: HashMap<usize, usize> = HashMap::new();
+    let mut user_data_indexes: HashSet<usize> = HashSet::new();
+
+    if !expecting_data {
+        for (pos, par) in parameters.c_parameters.iter().enumerate() {
+            if async && par.c_type == "gpointer" {
```

Comment:
There might be other parameters apart from the `user_data` that are `gpointer`. E.g. the `user_data` for a normal callback, or something arbitrary (at least in the latter case we don't support that and should ignore the whole function instead)

Reply:
I think it's a minority and for now I never encountered it. I can add a note for now if you want though?

Reply:
Why don't you use the `closure` index from the annotation here to prevent such problems to begin with? You already have the information available

Reply:
I don't trust gir much and indexes even less.

Reply:
We'll notice if they're wrong, stuff elsewhere would fail already then. By using the indexes here this code could be come more robust at least for the other cases.

Reply:
For now, I think it's preferrable to focus on what we're sure and then as a second step we can always extend it. I think it's reasonable?

Reply:
Add a FIXME comment then

Reply:
Done.